pkg/cmd/cli/secrets/source.go

```diff
@@ -0,0 +1,198 @@
+package secrets
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	CreateSourceSecretRecommendedCommandName = "new-sourcecfg"
+
+	createSourceSecretLong = `
+Create a new source secret
+
+Source secrets are used to authenticate against SCM servers.
+
+When creating applications, you may have a SCM server that requires authentication. In order for the
+nodes to clone source code on your behalf, they have to have the credentials. You can provide this
+information by creating a sourcecfg secret and attaching it to your service account.
+
+If your source authentication method requires only username and password or token, add it by using:
+
+  $ %[1]s SECRET_NAME --username=USERNAME --password=PASSWORD
+
+If your source authentication method requires private ssh key, add it by using:
+
+  $ %[1]s SECRET_NAME --ssh-private-key=SSH_KEY
+
+If your source authentication method requires a custom CA certificate, add it by using:
+
+  $ %[1]s SECRET_NAME --ca-cert=CERTIFICATE
+
+If you do already have a .gitconfig file needed for authentication, you can create a gitconfig secret by using:
+
+  $ %[2]s SECRET_NAME path/to/.gitconfig`
+)
+
+type CreateSourceSecretOptions struct {
+	SecretNamespace string
+	SecretName      string
+	Username        string
+	Password        string
+	PrivateKeyPath  string
+	SourceURL       string
+	CertificatePath string
+	GitConfigPath   string
+
+	PromptForPassword bool
+
+	Reader io.Reader
+	Out    io.Writer
+
+	SecretsInterface client.SecretsInterface
+}
+
+const (
+	SecretTypeSource api.SecretType = "Source"
+)
+
+func NewCmdCreateSourceSecret(name, fullName string, f *kcmdutil.Factory, reader io.Reader, out io.Writer, newSecretFullName, ocEditFullName string) *cobra.Command {
+	o := &CreateSourceSecretOptions{
+		Out:    out,
+		Reader: reader,
+	}
+
+	cmd := &cobra.Command{
+		Use:   fmt.Sprintf("%s SECRET_NAME --username=USERNAME --password=PASSWORD --ca-cert=CERTIFICATE --ssh-private-key=SSH_KEY --gitconfig=GITCONFIG", name),
```

Comment:
Use `FILENAME` here as well for flags that take the path to a file. If `--gitconfig` can be (optionally) tacked on to anything then I'd say the usage line is:

```
%s SECRET_NAME (--username=USERNAME --password=PASSWORD --ca-cert=FILENAME | --ssh-private-key=FILENAME) [--gitconfig=FILENAME]
```

which translates to something like "one of username+pass+ca-cert **or** ssh-private-key **required**, **plus optionally** a gitconfig".